### PR TITLE
Initialize EnsembleConfig only using from_dict() pathway

### LINKED ISSUE #4048: Simplify EnsembleConfig

### PR DESCRIPTION
**Pull request by DanSava**
**Issue**
Resolves #4048

**Approach**
Initialize EnsembleConfig only using from_dict() pathway

## Pre review checklist

- [ ] Added appropriate release note label
- [ ] PR title captures the intent of the changes, and is fitting for release notes.
- [ ] Commit history is consistent and clean, in line with the [contribution guidelines](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md).

Adding labels helps the maintainers when writing release notes. This is the [list of release note labels](https://github.com/equinor/ert/labels?q=release-notes).

### CODE REVIEW

**Review by eivindjahren — COMMENTED** [review]
Seems to break something on `semeio`:
```
The above exception was the direct cause of the following exception:

    @pytest.mark.usefixtures("setup_poly_ert")
    def test_localisation_field2():
        # Make a 3D grid with some inactive cells
        grid_filename = "grid3D.EGRID"
        grid = create_box_grid_with_inactive_and_active_cells(grid_filename)
    
        # Make some field parameters and some scalingfactor parameters
        poly_config_file = "poly.ert"
        create_field_and_scaling_param_and_update_poly_ert(
            poly_config_file, grid_filename, grid
        )
    
        # Create a discrete parameter to represent a region parameter
        segment_filename1 = "Region1.GRDECL"
        create_region_parameter(segment_filename1, grid)
        segment_filename2 = "Region2.GRDECL"
        create_region_parameter(segment_filename2, grid)
    
>       ert = LibresFacade.from_config_file("poly.ert")

/home/runner/work/ert/ert/semeio/tests/jobs/localisation/test_integration.py:412: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/opt/hostedtoolcache/Python/3.9.15/x64/lib/python3.9/site-packages/ert/libres_facade.py:615: in from_config_file
    return cls(EnKFMain(ResConfig(config_file), read_only))
/opt/hostedtoolcache/Python/3.9.15/x64/lib/python3.9/site-packages/ert/_c_wrappers/enkf/res_config.py:97: in __init__
    self._alloc_from_content(
/opt/hostedtoolcache/Python/3.9.15/x64/lib/python3.9/site-packages/ert/_c_wrappers/enkf/res_config.py:304: in _alloc_from_content
    self.ensemble_config = EnsembleConfig.from_dict(config_dict=config_content_dict)
/opt/hostedtoolcache/Python/3.9.15/x64/lib/python3.9/site-packages/ert/_c_wrappers/enkf/ensemble_config.py:327: in from_dict
    field_node = cls.get_field_node(
/opt/hostedtoolcache/Python/3.9.15/x64/lib/python3.9/site-packages/ert/_c_wrappers/enkf/ensemble_config.py:258: in get_field_node
    return EnkfConfigNode.create_field(
/opt/hostedtoolcache/Python/3.9.15/x64/lib/python3.9/site-packages/ert/_c_wrappers/enkf/config/enkf_config_node.py:342: in create_field
    config_node._update_parameter_field(
```

**Comment by DanSava** [comment]
This PR does not entirely resolve https://github.com/equinor/ert/issues/4048. 
The following issues still remain:

*  `EnsembleConfig` still inherits from `BaseCClass` not really possible to avoid it currently because the `ensemble_config_type` is used by [update.cpp, enkf_state.cpp, obs_vector.cpp, enkf_fs.cpp, enkf_obs.cpp, misfit_ensemble.cpp] and untangling all the spaghetti code in one PR I don't think it is healthy.

*  `EnkfConfigNode` needs to be lifted into python which is another can of really tangled spaghetti but should be attacked before the above point.

**Review by oysteoh — COMMENTED** [review]
From my point of view this looks good to me ! But i'll leave it up to @eivindjahren  to approve since he have had comments in the PR :+1:

**Review by eivindjahren — APPROVED** [review]
LGTM!

File: src/ert/_c_wrappers/enkf/ensemble_config.py
Comment (valentin-krasontovitsch) [reply src/ert/_c_wrappers/enkf/ensemble_config.py:310]: yes indeed - we don't have good coverage here yet, as the config dict generator doesn't cover the schedule key.

changing the config file `test-data/configuration_tests/ensemble_config.ert`, to include the line:

```
SCHEDULE_PREDICTION_FILE input/schedule.sch PARAMETER:input/param
```

and correspondingly updating the config dict in `tests/unit_tests/c_wrappers/res/enkf/test_ensemble_config.py`, line 95, to read

```
                   ConfigKeys.PARAMETER_KEY: "input/param",
```

gives us... drum roll... a util abort : )

File: src/clib/old_tests/enkf/test_enkf_gen_data_config_parse.cpp
Comment (valentin-krasontovitsch) [inline src/clib/old_tests/enkf/test_enkf_gen_data_config_parse.cpp:98]: yaaa now I see where these test cases came from! so it shouldn't be much work to steal the strings from the comments : )

**Review by valentin-krasontovitsch — CHANGES_REQUESTED** [review]
Looking good, it's a massive amount of work, so thank you for tackling it. I do have a few comments... : )

File: src/ert/_c_wrappers/enkf/ensemble_config.py
Comment (DanSava) [reply src/ert/_c_wrappers/enkf/ensemble_config.py:147]: This was the behaviour before the changes so I thought I would keep it as it was. If I throw an error now that will probably break some existing config files and I wanted to avoid that.

File: src/ert/_c_wrappers/enkf/ensemble_config.py
Comment (DanSava) [reply src/ert/_c_wrappers/enkf/ensemble_config.py:186]: You added the line where it mentions that the only valid value is ASCII, was that a mistake?  Maybe I need to only allow INPUT_FORMAT to be "ASCII"?

File: tests/unit_tests/c_wrappers/res/config/test_config_parser.py
Comment (DanSava) [reply tests/unit_tests/c_wrappers/res/config/test_config_parser.py:335]: > i've read at some point that one shouldn't necessarily test implementation details, or every method / function that one writes. it was argued that this kind of test might be nice to have during development (in particular employing TDD), to make sure that the function turns out alright, but can be removed afterwards.

I personally did not read this at any point, and are you arguing now that simple functionality unit tests can/should be removed? Or am I missing the point? 

> when / if we ever change the implementation of ensemble config, or how the options are parsed, this test will have to be rewritten / updated as well, and that's an unnecessary slow down, some would say.

Those that say "that's an unnecessary slow down" are focusing too much on writing code just for the sake of writing a lot of code, not really focusing on catching errors before they hit the users.

>one might be able to make sure that this parsing works correctly by supplying enough cases via a config file / dict, using the hypothesis approach - and perhaps write a few non-correct examples (again on the level of config dict / file), to make sure that it errors / ignores in the expected way.

I agree that the test can be improved and should also account for more edge case like when the list provided is empty or if the list has non-string elements but in those cases, I assumed that would result in an invalid config file, and the parsing of the config file should handle it. But I will make a not to improve the test.

File: tests/unit_tests/c_wrappers/res/enkf/test_ensemble_config.py
Comment (DanSava) [reply tests/unit_tests/c_wrappers/res/enkf/test_ensemble_config.py:191]: How about a comment ?

File: src/ert/_c_wrappers/enkf/ensemble_config.py
Comment (DanSava) [reply src/ert/_c_wrappers/enkf/ensemble_config.py:92]: Just preserving some documentation from the C code that was removed

File: src/ert/_c_wrappers/enkf/ensemble_config.py
Comment (DanSava) [reply src/ert/_c_wrappers/enkf/ensemble_config.py:117]: I should consider adding type hints good suggestion

File: src/ert/_c_wrappers/enkf/ensemble_config.py
Comment (DanSava) [reply src/ert/_c_wrappers/enkf/ensemble_config.py:310]: Did you also create that `param` file in `input/param` ? Also are you sure that it needs to have a particular format?

File: src/ert/_c_wrappers/enkf/ensemble_config.py
Comment (valentin-krasontovitsch) [reply src/ert/_c_wrappers/enkf/ensemble_config.py:186]: did I? XD

then I must have had a reason... let me check real quick...

File: src/ert/_c_wrappers/enkf/ensemble_config.py
Comment (valentin-krasontovitsch) [reply src/ert/_c_wrappers/enkf/ensemble_config.py:310]: yeah i now checked on main and saw that the util abort was that the file is missing. maybe that was the case on your branch as well, because no, i didn't create it. let me try again!

Comment (valentin-krasontovitsch) [reply src/ert/_c_wrappers/enkf/ensemble_config.py:310]: so creating the file gives an exception, raised by the test, as it cannot create the representation of the parameter file:

```
/home/val/code/equinor/ert/src/ert/_c_wrappers/enkf/config/gen_kw_config.py:82: in getParameterFile:
...
E       UnicodeDecodeError: 'utf-8' codec can't decode byte 0xaf in position 0: invalid start byte
```

trying to fill the file with possibly meaningful content, like

```
WIGGLYNESS UNIFORM 0.98 1.03
```

gives the same error. this is on d0f4143fd, by the way, before your merged PR.

still trying to figure out if the parameter thingy is actually used in the context of a schedule prediction file. it seems to me like this was broken anyways...

Comment (valentin-krasontovitsch) [reply src/ert/_c_wrappers/enkf/ensemble_config.py:310]: on current main, adding the changes described above, leads to the following value of `schedule_file_list` in ensemble config's `from_dict`:

```
[
  '/tmp/pytest-of-val/pytest-65/test_ensemble_config_construct0/test_data/input/schedule.sch', 
  'PARAMETER:input/param'
]
```
but then in `__init__` it tries to add each of these as a separate schedule file thingy, and it crashed on the second one, because it interprets the whole string as the first argument to schedule, where it expects a template file:

```
Error message: gen_kw_config_set_template_file: the template_file:PARAMETER:input/param does not exist - aborting.
```

so the handling of that schedule config thingy needs to either be covered by moar tests (all the tests wohoo) and fixed, or canned, because it's not necessary to have the other args (maybe we can ask a schedule prediction file expert tomorrow)

File: src/ert/_c_wrappers/enkf/ensemble_config.py
Comment (valentin-krasontovitsch) [reply src/ert/_c_wrappers/enkf/ensemble_config.py:334]: XD the classic rebase problem - do i actually consider if the changes i'm rebasing on effect my code, or do i just pray for no merge conflicts / resolve conflicts, and then everything should be fine ; )

File: tests/unit_tests/c_wrappers/res/enkf/test_ensemble_config.py
Comment (valentin-krasontovitsch) [reply tests/unit_tests/c_wrappers/res/enkf/test_ensemble_config.py:191]: comment is good, but i was thinking an extra part of the pytest param thingy, then it ... i dunno, shows up immediately when the test fails? but comment will definitely be sufficient as well!

after a quick google search, i think i found what i was thinking of, sorry for not being more explicit before: there's this `id` keyword you can use in parameters with pytest, see the [docs here](https://docs.pytest.org/en/7.1.x/example/parametrize.html#different-options-for-test-ids)

i'll leave it up to you whether you prefer comments or IDs!

File: tests/unit_tests/c_wrappers/res/enkf/test_ensemble_config.py
Comment (valentin-krasontovitsch) [reply tests/unit_tests/c_wrappers/res/enkf/test_ensemble_config.py:181]: ok I'm confused, but also I think I see now - that string is not THE gen data key, it is just A key that is used in the gen data context? like in the docs:

```
GEN_DATA 4DWOC  INPUT_FORMAT:ASCII   RESULT_FILE:SimulatedWOC%d.txt   REPORT_STEPS:10,100
```

the `GEN_DATA_KEY` from the tests is the same as `4WDOC` from the example, eh? I suggest using carefully chosen words from more or less popular monty python's works, it seems to be the pythonic way ; )

File: tests/unit_tests/c_wrappers/res/config/test_config_parser.py
Comment (valentin-krasontovitsch) [reply tests/unit_tests/c_wrappers/res/config/test_config_parser.py:335]: ok i remember now that i heard it, it was a talk eivind linked to on slack - https://equinor.slack.com/archives/C3HLP8ZTQ/p1659420684498539 [internal link], the talk is https://youtu.be/EZ05e7EMOLM - there's a lot of introduction / background in the beginning, which might still be worth a listen, but the main message i'm referring to start at around timestamp 24:00

> are you arguing now that simple functionality unit tests can/should be removed?

yeah i know it sounds crazy. but my point (also with the "higher level approach") is this. if we have tests that pass in config files or dict with contents that cover the cases you have in the unit tests (by providing the necessary lines / key-values that will hit exactly those functions the unit tests are testing with the right arguments), then we get the same coverage - but we don't need to worry about how we implemented it, and are free to change details about those small methods later, or refactor the way its implemented (as in for instance use some other methods to achieve the same thing), without having to change our tests.

i hope this makes it a bit clearer, if not, we can disregard it here, and i will try to bring it up in a retrospective instead, or some other forum.

File: src/ert/_c_wrappers/enkf/ensemble_config.py
Comment (valentin-krasontovitsch) [reply src/ert/_c_wrappers/enkf/ensemble_config.py:186]: i mean look at this beautiful function:

```
gen_data_config_type *
gen_data_config_alloc_GEN_DATA_result(const char *key,
                                      gen_data_file_format_type input_format) {
    gen_data_config_type *config = gen_data_config_alloc(key, true);
                      
    if (input_format == ASCII_TEMPLATE)
        util_abort("%s: Sorry can not use INPUT_FORMAT:ASCII_TEMPLATE\n",
                   __func__);

    if (input_format == GEN_DATA_UNDEFINED)
        util_abort("%s: Sorry must specify valid values for input format.\n",
                   __func__);

    config->input_format = input_format;
    return config;
}
```

so much util abort ^^

also there's a comment further down:

```
/**                                                                          
   The valid options are:    

   INPUT_FORMAT:(ASCII|ASCII_TEMPLATE)  
   RESULT_FILE:<filename to read EnKF <== Forward model>
 
*/
```

i think i forgot to fix (remove!?) that comment, but the point is, any allocation in our code of a gen data config thingy happens through `gen_data_config_alloc_GEN_DATA_result` (even `..._alloc_full` calls that one), and you can see what the alloc result function does :sweat_smile: 

so I guess that's why i made that change in the docs...

god i wrote an essay in the commit message, but no mention of the input format thing :see_no_evil:

File: tests/unit_tests/c_wrappers/res/config/test_config_parser.py
Comment (valentin-krasontovitsch) [reply tests/unit_tests/c_wrappers/res/config/test_config_parser.py:335]: btw i was also kinda shocked when i was presented with the notion to stop writing this kind of unit tests, as that was how i was literally introduced to TDD... and it seems so basic and logical, to get the safety of (trying to) covering each method you have, then you gotta be good, right? so it seems nonsensical when somebody says never mind, we don't need those...

File: src/ert/_c_wrappers/enkf/ensemble_config.py
Comment (valentin-krasontovitsch) [reply src/ert/_c_wrappers/enkf/ensemble_config.py:92]: great! maybe we can get back to it during validation work.

File: src/ert/_c_wrappers/enkf/ensemble_config.py
Comment (valentin-krasontovitsch) [inline src/ert/_c_wrappers/enkf/ensemble_config.py:298]: is it possible that this one can be a static method as well? doesn't look like cls is actually used. sorry didn't catch it first time around

File: tests/unit_tests/c_wrappers/res/config/test_config_parser.py
Comment (eivindjahren) [reply tests/unit_tests/c_wrappers/res/config/test_config_parser.py:335]: Thing is that one persons implementation detail is someone elses feature. Just because this function is private to ensemble_config, it is still valid to test it in that domain, but it would be good to test it through a public interface. However, I feel this is a good time to instead use doctest as the problem here is that these tests are weird when out of context. I would much rather have these as part of the docstring, and then testsed through `pytest --doctest-modules`.

Ie 

```
def _str_to_bool(text):
  """
  This function converts truthy text to boolean values according to the 
  rules of the FORWARD_INIT keyword, we keep these rules for backwards compatability

  First, any upper/lower case true/false value is converted to the corresponding boolean value:
  
  >>> _str_to_bool("TRUE")
  True
  >>> _str_to_bool("true")
  True
  
  ...
```

Comment (eivindjahren) [reply tests/unit_tests/c_wrappers/res/config/test_config_parser.py:335]: Separately: is logging the failure the correct behavior here? We have this `ConfigValidationError` we raise to tell the user that there was an error in the configuration file now.

Comment (valentin-krasontovitsch) [reply tests/unit_tests/c_wrappers/res/config/test_config_parser.py:335]: ohhh doctests time to learn something new :heart_eyes:

Comment (eivindjahren) [reply tests/unit_tests/c_wrappers/res/config/test_config_parser.py:335]: Work on adding doctest is done here: https://github.com/equinor/ert/pull/4278

File: src/ert/_c_wrappers/enkf/ensemble_config.py
Comment (DanSava) [reply src/ert/_c_wrappers/enkf/ensemble_config.py:273]: Fair point, I will remove the numbers

File: src/ert/_c_wrappers/enkf/ensemble_config.py
Comment (DanSava) [reply src/ert/_c_wrappers/enkf/ensemble_config.py:310]: I will fix it for now and update the test `test_ensemble_config_constructor` to cover the case where a parameter file is provided. 

I created an issue (https://github.com/equinor/ert/issues/4288) to investigate if the extra arguments should be allowed or not.

Comment (valentin-krasontovitsch) [reply src/ert/_c_wrappers/enkf/ensemble_config.py:310]: great, thank you!

File: src/ert/_c_wrappers/enkf/ensemble_config.py
Comment (valentin-krasontovitsch) [reply src/ert/_c_wrappers/enkf/ensemble_config.py:334]: did you get some new information here? the code doesn't seem to have changed in the new PR :thinking:

File: src/ert/_c_wrappers/enkf/ensemble_config.py
Comment (valentin-krasontovitsch) [reply src/ert/_c_wrappers/enkf/ensemble_config.py:337]: this question still remains open, as for summary list

File: src/ert/_c_wrappers/enkf/config/enkf_config_node.py
Comment (valentin-krasontovitsch) [reply src/ert/_c_wrappers/enkf/config/enkf_config_node.py:254]: i guess the reasoning is as above, you're keeping the behaviour we had before, eh? sounds fair.

File: src/ert/_c_wrappers/enkf/ensemble_config.py
Comment (valentin-krasontovitsch) [reply src/ert/_c_wrappers/enkf/ensemble_config.py:147]: sounds very fair indeed!

File: src/ert/_c_wrappers/enkf/ensemble_config.py
Comment (valentin-krasontovitsch) [reply src/ert/_c_wrappers/enkf/ensemble_config.py:334]: perfect, thanks!

File: src/ert/_c_wrappers/enkf/ensemble_config.py
Comment (valentin-krasontovitsch) [reply src/ert/_c_wrappers/enkf/ensemble_config.py:337]: perfect, thanks!

**Pull request merged**
